### PR TITLE
Setup cache tests

### PR DESCRIPTION
**Pull request by devongovett**
This sets up some tests for the cache to ensure it works as expected when changing various things between builds. So far I've set up a couple basic tests, and stubbed out some scenarios that we should cover. Will be adding more as well. Please let me know if you think of scenarios not yet covered. 😄

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 7.00s | +100.00ms |
| Cached | 2.60s | +108.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 256.00ms | -42.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 256.00ms | -42.00ms 🚀 |
| dist/legacy/index.42516b4a.js | 1.10kb | +0.00b | 1.72s | -196.00ms 🚀 |
| dist/modern/index.8a1ced31.js | 1.10kb | +0.00b | 1.71s | -196.00ms 🚀 |
| dist/legacy/index.html | 701.00b | +0.00b | 616.00ms | -82.00ms 🚀 |
| dist/modern/index.html | 701.00b | +0.00b | 612.00ms | -83.00ms 🚀 |
| dist/legacy/index.0e158b6b.css | 77.00b | +0.00b | 1.71s | -197.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 213.00ms | -45.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 212.00ms | -46.00ms 🚀 |
| dist/legacy/index.42516b4a.js | 1.10kb | +0.00b | 470.00ms | -68.00ms 🚀 |
| dist/modern/index.8a1ced31.js | 1.10kb | +0.00b | 469.00ms | -68.00ms 🚀 |
| dist/legacy/index.html | 701.00b | +0.00b | 472.00ms | -69.00ms 🚀 |
| dist/modern/index.html | 701.00b | +0.00b | 469.00ms | -69.00ms 🚀 |
| dist/legacy/index.0e158b6b.css | 77.00b | +0.00b | 470.00ms | -69.00ms 🚀 |
| dist/modern/index.e480862d.css | 77.00b | +0.00b | 469.00ms | -68.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 27.71s | +287.00ms |
| Cached | 2.32s | +36.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 155.00ms | +10.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 478.42kb | +0.00b | 208.00ms | -163.00ms 🚀 |
| dist/PermalinkedComment.c85d137d.js | 4.22kb | +0.00b | 206.00ms | -163.00ms 🚀 |
| dist/UserProfile.30e15baf.js | 1.62kb | +0.00b | 206.00ms | -163.00ms 🚀 |
| dist/NotFound.5aa6b041.js | 387.00b | +0.00b | 206.00ms | -163.00ms 🚀 |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 164.00ms | -147.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.84m | +838.00ms |
| Cached | 4.95s | +93.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/EmojiPickerComponent.bab709db.js | 97.89kb | +0.00b | 13.66s | +1.52s ⚠️ |
| dist/Toolbar.33a83ae3.js | 92.34kb | +0.00b | 57.85s | +2.99s ⚠️ |
| dist/media-viewer.a8f3a476.js | 72.02kb | +0.00b | 17.35s | +1.47s ⚠️ |
| dist/card.ce67dc35.js | 55.09kb | +0.00b | 15.14s | +2.13s ⚠️ |
| dist/Modal.e3bd77d0.js | 41.40kb | +0.00b | 6.03s | +652.00ms ⚠️ |
| dist/component.59edff1f.js | 30.78kb | +0.00b | 3.22s | +932.00ms ⚠️ |
| dist/component.37867f50.js | 22.51kb | +0.00b | 6.03s | +654.00ms ⚠️ |
| dist/DatePicker.cbe44a9b.js | 21.17kb | +0.00b | 11.34s | +1.38s ⚠️ |
| dist/EmojiPickerComponent.898d8be6.js | 18.23kb | +0.00b | 11.34s | +1.37s ⚠️ |
| dist/smartMediaEditor.c0ff7bfa.js | 16.87kb | +0.00b | 56.64s | +2.83s ⚠️ |
| dist/js.48be70a4.js | 16.40kb | +0.00b | 6.03s | +654.00ms ⚠️ |
| dist/ui.cee430a3.js | 13.97kb | +0.00b | 14.56s | +1.60s ⚠️ |
| dist/workerHasher.65b8a1b7.js | 11.78kb | +0.00b | 14.56s | -38.80s 🚀 |
| dist/component.11476a92.js | 6.05kb | +0.00b | 6.43s | +1.05s ⚠️ |
| dist/card.4ac5c1b2.js | 5.61kb | +0.00b | 17.34s | +3.53s ⚠️ |
| dist/media-viewer.bc661bce.js | 3.92kb | +0.00b | 17.35s | +1.49s ⚠️ |
| dist/EmojiPickerComponent.624bf3f4.js | 3.58kb | +0.00b | 11.33s | +1.37s ⚠️ |
| dist/index.393afe59.css | 3.46kb | +0.00b | 6.23s | -1.52m 🚀 |
| dist/png-chunks-extract.8d15a914.js | 3.41kb | +0.00b | 6.03s | +654.00ms ⚠️ |
| dist/dropzone.2f1a1e17.js | 3.23kb | +0.00b | 52.82s | +4.14s ⚠️ |
| dist/Modal.2835744c.js | 3.08kb | +0.00b | 6.03s | +1.27s ⚠️ |
| dist/16.d1b8e6f7.js | 2.38kb | +0.00b | 6.53s | +4.27s ⚠️ |
| dist/card.cadccd96.js | 2.03kb | +0.00b | 31.18s | +15.14s ⚠️ |
| dist/ResourcedEmojiComponent.d9a49606.js | 2.00kb | +0.00b | 6.43s | +1.98s ⚠️ |
| dist/date.d49e64f8.js | 1.73kb | +0.00b | 6.96s | +964.00ms ⚠️ |
| dist/16.f40c1329.js | 1.73kb | +0.00b | 6.65s | +935.00ms ⚠️ |
| dist/images.80270b1f.js | 1.67kb | +0.00b | 6.96s | +741.00ms ⚠️ |
| dist/16.86a8c7d4.js | 1.66kb | +0.00b | 6.43s | -1.23s 🚀 |
| dist/feedback.8b7c7928.js | 1.64kb | +0.00b | 7.46s | +459.00ms ⚠️ |
| dist/list-number.4632b459.js | 1.46kb | +0.00b | 7.08s | +602.00ms ⚠️ |
| dist/status.cccf04be.js | 1.46kb | +0.00b | 7.31s | +461.00ms ⚠️ |
| dist/16.cff410ff.js | 1.38kb | +0.00b | 9.85s | +4.31s ⚠️ |
| dist/code.af9ca2b3.js | 1.38kb | +0.00b | 6.83s | +832.00ms ⚠️ |
| dist/16.c1ed07f8.js | 1.33kb | +0.00b | 6.43s | -1.22s 🚀 |
| dist/16.6603f2f3.js | 1.33kb | +0.00b | 9.56s | +4.01s ⚠️ |
| dist/16.a4b392c2.js | 1.32kb | +0.00b | 3.20s | +946.00ms ⚠️ |
| dist/16.249c644a.js | 1.31kb | +0.00b | 6.84s | +842.00ms ⚠️ |
| dist/link.d12cc130.js | 1.30kb | +0.00b | 7.08s | +602.00ms ⚠️ |
| dist/heading6.4c4017db.js | 1.30kb | +0.00b | 7.46s | +458.00ms ⚠️ |
| dist/heading3.93606b9c.js | 1.29kb | +0.00b | 7.46s | +616.00ms ⚠️ |
| dist/16.ab5c0088.js | 1.27kb | +0.00b | 6.84s | +1.12s ⚠️ |
| dist/emoji.e603e70e.js | 1.23kb | +0.00b | 6.96s | +740.00ms ⚠️ |
| dist/16.afee1494.js | 1.23kb | +0.00b | 6.65s | +935.00ms ⚠️ |
| dist/16.e60b4d77.js | 1.21kb | +0.00b | 3.20s | +945.00ms ⚠️ |
| dist/16.7c7a40c9.js | 1.19kb | +0.00b | 6.65s | +935.00ms ⚠️ |
| dist/16.06d5a67b.js | 1.18kb | +0.00b | 6.84s | +842.00ms ⚠️ |
| dist/heading5.dde15cec.js | 1.17kb | +0.00b | 7.46s | +458.00ms ⚠️ |
| dist/expand.6889aead.js | 1.16kb | +0.00b | 11.34s | +4.33s ⚠️ |
| dist/media-card-analytics-error-boundary.6a25b970.js | 1.15kb | +0.00b | 50.72s | +22.04s ⚠️ |
| dist/16.2096b9f3.js | 1.14kb | +0.00b | 3.20s | +945.00ms ⚠️ |
| dist/16.f040d62d.js | 1.14kb | +0.00b | 6.65s | +1.11s ⚠️ |
| dist/16.ecf8a702.js | 1.13kb | +0.00b | 6.43s | -1.69s 🚀 |
| dist/16.1b342686.js | 1.13kb | +0.00b | 9.85s | +4.31s ⚠️ |
| dist/16.2e7f52c6.js | 1.13kb | +0.00b | 6.65s | +1.11s ⚠️ |
| dist/16.da43aa07.js | 1.13kb | +0.00b | 6.43s | -1.23s 🚀 |
| dist/heading2.e219309a.js | 1.11kb | +0.00b | 7.31s | +461.00ms ⚠️ |
| dist/16.dd07505e.js | 1.10kb | +0.00b | 3.20s | -1.24s 🚀 |
| dist/component.40194190.js | 1.07kb | +0.00b | 6.43s | +1.98s ⚠️ |
| dist/mention.b1828447.js | 1.07kb | +0.00b | 7.08s | +602.00ms ⚠️ |
| dist/heading4.5316f6f2.js | 1.06kb | +0.00b | 7.46s | +458.00ms ⚠️ |
| dist/16.317b9164.js | 1.05kb | +0.00b | 9.94s | +4.23s ⚠️ |
| dist/layout.ac6734f9.js | 1.05kb | +0.00b | 7.08s | +862.00ms ⚠️ |
| dist/divider.d2be55a1.js | 1.03kb | +0.00b | 6.96s | +740.00ms ⚠️ |
| dist/quote.f4e09143.js | 1.02kb | +0.00b | 7.31s | +661.00ms ⚠️ |
| dist/action.cb287eb0.js | 1.00kb | +0.00b | 6.84s | +842.00ms ⚠️ |
| dist/decision.91f3af14.js | 1010.00b | +0.00b | 6.96s | +741.00ms ⚠️ |
| dist/panel-warning.68b04096.js | 1008.00b | +0.00b | 7.18s | +535.00ms ⚠️ |
| dist/list.057cb664.js | 978.00b | +0.00b | 7.08s | +602.00ms ⚠️ |
| dist/heading1.fa66b375.js | 976.00b | +0.00b | 7.31s | +461.00ms ⚠️ |
| dist/panel-error.c4b05f6f.js | 906.00b | +0.00b | 7.18s | +704.00ms ⚠️ |
| dist/panel.9dbc048e.js | 897.00b | +0.00b | 7.18s | +535.00ms ⚠️ |
| dist/table.b35545be.js | 889.00b | +0.00b | 7.31s | +461.00ms ⚠️ |
| dist/panel-success.7eb0809a.js | 845.00b | +0.00b | 7.18s | +535.00ms ⚠️ |
| dist/panel-note.ec8386b7.js | 841.00b | +0.00b | 7.18s | +535.00ms ⚠️ |
| dist/media-picker-analytics-error-boundary.553055f6.js | 837.00b | +0.00b | 53.73s | +2.74s ⚠️ |
| dist/media-card-analytics-error-boundary.964206cc.js | 833.00b | +0.00b | 15.82s | +2.89s ⚠️ |
| dist/media-viewer-analytics-error-boundary.bbe40a02.js | 757.00b | +0.00b | 17.52s | +1.63s ⚠️ |
| dist/EmojiPickerComponent.a968dafe.js | 535.00b | +0.00b | 55.85s | +45.88s ⚠️ |
| dist/simpleHasher.f8e69464.js | 528.00b | +0.00b | 14.56s | +1.63s ⚠️ |
| dist/simpleHasher.32d9352e.js | 528.00b | +0.00b | 56.64s | +2.84s ⚠️ |
| dist/index.html | 119.00b | +0.00b | 1.78s | +1.10s ⚠️ |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/272721837246382605)**

**Comment by devongovett** [comment]
Let's get this setup in as a baseline and expand the missing tests in separate PRs.

**Comment by mischnic** [comment]
The watcher tests are failing with `Uncaught Cannot find parcel plugin "@parcel/reporter-cli"`

File: packages/core/integration-tests/test/cache.js [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+  try {
+    await fn();
+  } catch (err) {
+    threw = true;
+    e = err;
+  }
+
+  assert(threw, 'did not throw');
+  return e;
+}
```
Comment (mischnic) [inline packages/core/integration-tests/test/cache.js:50]: `await assert.throws(() => ...)`?

File: packages/core/integration-tests/test/cache.js [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+    let e = await assertThrows(async () => {
+      await runBundle();
+    });
+    assert.equal(
+      e.message,
+      "Failed to resolve './nested/test' from './src/index.js'",
+    );
+  });
+
+  describe('babel', function() {
```
Comment (padmaia) [inline packages/core/integration-tests/test/cache.js:115]: What do you think of the idea of moving plugin specific tests to the plugin packages? The advantages I see are that tests would be easier to find and easier to clean up if we migrate to a different plugin in the default config. It's not something I feel super strong about and it's definitely not a blocker, just something I've been thinking about.

**Review by padmaia — APPROVED** [review]

**Pull request merged**